### PR TITLE
Fix ability to import from pages, API routes, and queries/mutations

### PR DESCRIPTION
**Pull request by Skn0tt**
Closes: #1013
Closes: #734

### What are the changes and their implications?

Blitz's file pipeline moves files from *A* to *B*, but doesn't update any imports to *A*.
This PR makes sure that imports are rewritten, such that importing from any of the moved files doesn't break compilation.

### Checklist

- [X] Tests added for changes
- [ ] PR submitted to [blitzjs.com](https://github.com/blitz-js/blitzjs.com) for any user facing changes

### CODE REVIEW

**Comment by flybayer** [comment]
Still in progress?

**Comment by Skn0tt** [comment]
Yup

**Comment by Skn0tt** [comment]
So I was able to make this kind of work - you can now import stuff from `api/` or `pages/`, which at least fixes Quirrel's API for now.

There's one problem though: When importing from the frontend (see `examples/store/app/admin/pages/admin/products/index.tsx`), you'll see
```
TypeError: undefined is not a constructor (evaluating 'new _prisma_client__WEBPACK_IMPORTED_MODULE_0__["PrismaClient"]()')
```
And that makes sense! Since our `db/index.ts` has side-effects, webpack can't eliminate that from the frontend-bundle, so the browser (rightfully) fails at importing Prisma.

I see two options that we can do:

1. Forbid people from importing `api/` routes in the frontend. Seems a bit harsh.
2. Tell webpack that it can eliminate `db/index.ts` for the frontend bundle. Seems doable, but what about non-isomorphic imports with side-effects in userland?

Any other ideas @ryardley @flybayer @aem?

**Comment by flybayer** [comment]
I think it's fine to ensure anything in the `db` folder doesn't included in the frontend.

> but what about non-isomorphic imports with side-effects in userland?

What would be an example case for this?


Also, related issue to consider: https://github.com/blitz-js/blitz/issues/734

File: packages/server/src/with-blitz.ts [hunk 1]
```diff
@@ -24,6 +24,7 @@ export function withBlitz(nextConfig: any) {
         } else {
           config.module = config.module || {}
           config.module.rules = config.module.rules || []
+          config.module.rules.push({test: /db/, use: {loader: "null-loader"}})
```
Comment (Skn0tt) [inline packages/server/src/with-blitz.ts:27]: This makes sure that anything in the `db` folder is never included in the frontend bundle.

**Comment by Skn0tt** [comment]
> > but what about non-isomorphic imports with side-effects in userland?
> 
> What would be an example case for this?

I'm not sure tbh. That's a problem that existed before and now that I think about it, it's out of scope for this PR.

**Comment by Skn0tt** [comment]
https://github.com/blitz-js/blitz/pull/1456/commits/565752e2c9e7c9115123d9f3440dd0ca9e1fc111 closes https://github.com/blitz-js/blitz/issues/734.

File: packages/server/src/stages/rewrite-imports/index.ts [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+import path from "path"
+
+const isJavaScriptFile = (filepath: string) => filepath.match(/\.(ts|tsx|js|jsx)$/)
+
+const isInAppFolder = (s: string, cwd: string) => s.replace(cwd + path.sep, "").indexOf("app") === 0
+
+/**
+ * Returns a Stage that converts relative files paths to absolute
+ */
+export const createStageRewriteImports: Stage = ({config: {cwd}}) => {
```
Comment (Skn0tt) [inline packages/server/src/stages/rewrite-imports/index.ts:11]: This stage is very similar to https://github.com/blitz-js/blitz/blob/07b7a8a14809097d119c7b080a31abea10ca3d58/packages/server/src/stages/relative/index.ts, maybe we can refactor them to share code.

**Comment by Skn0tt** [comment]
> > but what about non-isomorphic imports with side-effects in userland?
> 
> What would be an example case for this?

I found one! The [current CI fail](https://github.com/blitz-js/blitz/pull/1456/checks?check_run_id=1402885086) seems to be caused by webpack trying to bundle `secure-password` in the frontend, which depends on Node APIs.

**Comment by Skn0tt** [comment]
It fails because of this line in `auth-utils.ts`:

```ts
const SP = new SecurePassword()
```

Because this is side-effectful, webpack tries to bundle it in the frontend.

We could easily fix this by changing the line like so:

```ts
const SP = typeof window === "undefined" ? new SecurePassword() : undefined;
```

But isn't that kind of ugly? How does Next.js handle this?

**Comment by Skn0tt** [comment]
Alright, so Next.js fails when importing Node APIs in the frontend: https://codesandbox.io/s/node-imports-test-z9wio
I guess that's the sensible way 😅

File: packages/server/src/stages/relative/pattern-relative-import.test.ts [hunk 3]
```diff
@@ -17,7 +17,7 @@ describe("patternRelativeImport", () => {
   })
 
   it("matches an import statement with no specifier", () => {
-    const input = `import from "./ding"`
+    const input = `import "./ding"`
```
Comment (Skn0tt) [inline packages/server/src/stages/relative/pattern-relative-import.test.ts:20]: `import from` isn't actual module syntax, so I changed this :D

File: packages/server/src/stages/rewrite-imports/index.ts [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+      ,
+      origin,
+      ,
+    ] = args as string[]
+
+    const parts = origin.split("/")
+
+    if (parts.includes("pages")) {
+      if (parts[0] === "app") {
+        parts.splice(0, 1)
```
Comment (Skn0tt) [inline packages/server/src/stages/rewrite-imports/index.ts:51]: If it's an import from a page, say from `app/pages/mypage.tsx`, we'll rewrite that import to `pages/mypage.tsx`.

File: packages/server/src/stages/rewrite-imports/index.ts [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+      }
+    }
+
+    if (parts.includes("api")) {
+      parts.splice(0, parts.indexOf("api"), "pages")
```
Comment (Skn0tt) [inline packages/server/src/stages/rewrite-imports/index.ts:56]: If it's an import from an API Route, say from `app/users/api/myRoute`, we'll rewrite it to `pages/api/myRoute`.

File: packages/server/src/stages/rewrite-imports/index.ts [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+    }
+
+    if (starImport) {
```
Comment (Skn0tt) [inline packages/server/src/stages/rewrite-imports/index.ts:59]: If it's a "star import" (what are they called irl?) like `import * as query from "app/queries/myQuery"`, I don't really know what to do. We'll just keep it as is for the moment.

In a perfect world, we'd somehow rewrite it to `app/_resolvers/queries/myQuery` but keep `query.default` as it is - I'm not so sure about how that's supposed to work, though.

File: packages/server/src/stages/rewrite-imports/index.ts [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+      // I don't know yet what to do
+      return args[0]
+    }
+
+    if (parts.includes("queries") || parts.includes("mutations")) {
```
Comment (Skn0tt) [inline packages/server/src/stages/rewrite-imports/index.ts:64]: Now the juicy part, which deals with #734. Whenever we see smth like `import { myFunc } from "app/queries/myQuery"`, we rewrite that to `app/_resolvers/queries/myQuery`.

When we see smth like `import myQuery, { myFunc } from "app/queries/myQuery"`, we rewrite it to the following two:
```ts
import myQuery from "app/queries/myQuery"
import { myFunc } from "app/_resolvers/queries/myQuery"
```

File: packages/server/src/stages/rewrite-imports/index.ts [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+      ,
+      ,
+      starImport,
+      defaultImportName,
+      combinedComma,
+      namedImportNames,
+      ,
+      origin,
+      ,
+    ] = args as string[]
```
Comment (ryardley) [inline packages/server/src/stages/rewrite-imports/index.ts:45]: Nice :) Now that's a few groups!

File: packages/server/src/stages/rewrite-imports/rewrite-imports.test.ts [hunk 9]
```diff
@@ -0,0 +1,213 @@
...
+}
+            `,
+          },
+        ],
+        expectedOutput: [
+          {
+            path: normalize("/projects/blitz/blitz/app/anyFile.ts"),
+            contents: `
+import query from "app/users/queries/getUser"
+import { someFunction } from "app/_resolvers/users/queries/getUser";
```
Comment (ryardley) [inline packages/server/src/stages/rewrite-imports/rewrite-imports.test.ts:203]: Should we be testing for dynamic imports too? `const bar = import('foo/queries/bar')`

File: packages/server/src/stages/relative/pattern-relative-import.test.ts (hunk 3 above)
Comment (ryardley) [reply packages/server/src/stages/relative/pattern-relative-import.test.ts:20]: Nice work

File: packages/server/src/stages/rewrite-imports/index.ts [hunk 10]
```diff
@@ -0,0 +1,85 @@
...
+    const newContents = replaceImports(contents)
+    file.contents = Buffer.from(newContents)
+
+    return file
+  })
+
+  return {stream}
+}
+
+export const patternImport = /(import\s+)(?:(?:(\*\s+as\s+\w+)|(\w+\s*)?(,\s*)?(?:(\{\s*\w+\s*(?:,\s*\w+\s*)*\}))?))(\s+from\s+["'])?([\w\.\\\/]+)(["']\)?)/gs
```
[the remark this answers is not in the document]
Comment (Skn0tt) [reply packages/server/src/stages/rewrite-imports/index.ts:31]: Yes it is: https://regex101.com/r/DO9kaT/4

Kinda proud that I got that working 😅

File: packages/server/src/stages/rewrite-imports/index.ts (hunk 8 above)
Comment (Skn0tt) [reply packages/server/src/stages/rewrite-imports/index.ts:45]: It is!

File: packages/server/src/stages/rewrite-imports/rewrite-imports.test.ts (hunk 9 above)
Comment (Skn0tt) [reply packages/server/src/stages/rewrite-imports/rewrite-imports.test.ts:203]: How do these work? Does `import("foo/queries/bar")` import the default export (the query)? Or is it like an `import *`?

Because if they work like `import *`, I'm not so sure as to how we can rewrite those in a sensible way...

File: packages/server/src/stages/rewrite-imports/rewrite-imports.test.ts [hunk 11]
```diff
@@ -0,0 +1,213 @@
...
+        "react",
+        '"',
+      ],
+      [`import db, {ProductCreateArgs} from "db"
+      type CreateProductInput = {
+        data: ProductCreateArgs["data"]
+      }`]: ["import ", , "db", ", ", "{ProductCreateArgs}", ' from "', "db", '"'],
+    }
+
+    Object.entries(successes).forEach(([input, expectedOutput]) => {
```
[the remark this answers is not in the document]
Comment (Skn0tt) [reply packages/server/src/stages/rewrite-imports/rewrite-imports.test.ts]: Makes sense ☺️ Done in https://github.com/blitz-js/blitz/pull/1456/commits/9c683cc4cf44499b8daab7f724d2bcd6ecedba19

File: packages/server/src/stages/rewrite-imports/rewrite-imports.test.ts (hunk 9 above)
Comment (ryardley) [reply packages/server/src/stages/rewrite-imports/rewrite-imports.test.ts:194]: > How do these work? Does import("foo/queries/bar") import the default export (the query)? Or is it like an import *?

I have no idea. Would have to look it up.

Comment (Skn0tt) [reply packages/server/src/stages/rewrite-imports/rewrite-imports.test.ts:194]: Tried it, it imports the whole module.
Screenshot 2020-11-16 at 11 10 08

If you have an idea about how to rewrite these, I'm all ears. Otherwise, I'd say we let it slide for the moment (I don't think that a lot of people will import these dynamically, anyway)

Comment (ryardley) [reply packages/server/src/stages/rewrite-imports/rewrite-imports.test.ts:194]: You can match the regex for it like this. 

```ts
/import\s?\(["']([^"']+)["']\);?/g
```

Comment (ryardley) [reply packages/server/src/stages/rewrite-imports/rewrite-imports.test.ts:194]: Might have some corner cases.

Comment (Skn0tt) [reply packages/server/src/stages/rewrite-imports/rewrite-imports.test.ts:194]: The problem is not detecting the import, but what to rewrite it to. See https://github.com/blitz-js/blitz/pull/1456#discussion_r523812784 for reference.

For imports from `api` and `pages` though, that's not a problem and I'll implement it.

File: packages/server/src/stages/rewrite-imports/index.ts (hunk 7 above)
Comment (Skn0tt) [reply packages/server/src/stages/rewrite-imports/index.ts:64]: I've also tried having the generated `queries/myQuery` file export everything from the resolvers: https://github.com/blitz-js/blitz/pull/1456/commits/aa67b5a3679dec03f510d2cf14be5b24b7b4c42a

This has the downside that it also bundles all transitive dependencies of the query, a lot of which cannot be tree-shaken because they have side-effects on the module level (even our default `auth-utils.ts` since it initialises `secure-password` on the top level).

One possible solution would be to make the additional export in https://github.com/blitz-js/blitz/pull/1456/commits/aa67b5a3679dec03f510d2cf14be5b24b7b4c42a only appear when there's actually some other file importing from there, but that would required coordination with the RPC stage.

File: packages/server/src/stages/rewrite-imports/rewrite-imports.test.ts (hunk 9 above)
Comment (ryardley) [reply packages/server/src/stages/rewrite-imports/rewrite-imports.test.ts:194]: Hmm sounds like we're almost getting into AST territory...

Comment (ryardley) [reply packages/server/src/stages/rewrite-imports/rewrite-imports.test.ts:194]: Still it can be done with regex.

Comment (ryardley) [reply packages/server/src/stages/rewrite-imports/rewrite-imports.test.ts:194]: Actually then there is the issue when you take a var and use it later 

```ts
const foo = import('thing/blat/boo');
//
foo.default()
```

Comment (ryardley) [reply packages/server/src/stages/rewrite-imports/rewrite-imports.test.ts:194]: Perhaps we should do a double whammy and use an ast to solve this as well as https://github.com/blitz-js/blitz/issues/734 ?

Comment (Skn0tt) [reply packages/server/src/stages/rewrite-imports/rewrite-imports.test.ts:194]: Could you help me understand what improvements an AST would bring in this case? Currently, I don't see a way to rewrite * imports and dynamic imports so that they work _at all_. What would you replace the code example above with?

I'll take a shot at the conditional exports mentioned in https://github.com/blitz-js/blitz/pull/1456#discussion_r524200987, I think that could be successfull.

Comment (Skn0tt) [reply packages/server/src/stages/rewrite-imports/rewrite-imports.test.ts]: Heureka ☺️ I guess I solved it with 15bb9cc. The `rpc` stage now generates both a `myQuery.ts` and a `myQuery.named.ts` that includes named exports.

In the rewrite-imports stage, imports that include any named imports are rewritten to use `myQuery.named.ts`.

**Comment by Skn0tt** [comment]
Anybody have an idea why 15bb9cc broke the Windows build?

**Comment by Skn0tt** [comment]
> Anybody have an idea why 15bb9cc broke the Windows build?

Hmm, seems like it's been something else.

**Comment by Skn0tt** [comment]
Anyway, seems like we've solved the main problem. Yay ☺️

**Comment by flybayer** [comment]
Hallelujah!! Beautiful work mate!

I did run into a snag, I refactored the auth example to import zod schemas from the mutation files but the build failed with this:

```
error - ../../../../../node_modules/node-gyp-build/index.js:1:0
Module not found: Can't resolve 'fs'
null
```

I pushed the commit with this change.

**Comment by Skn0tt** [comment]
That's because `mutations/login` imports `auth-utils`, which in turn imports `fs` via `secure-password`. Sadly, there's nothing we can do about that. Next.js would fail at that, too.

**Comment by flybayer** [comment]
Oh, I thought you fixed that somehow. So probably we want to add null-loaders for secure-password and maybe also for auth-utils.

**Comment by flybayer** [comment]
Another idea to consider separately would be to add ability to exclude a file from the client side bundle with a magic comment.

```
// blitz: server-only
```

**Comment by Skn0tt** [comment]
> we want to add null-loaders for secure-password

Another option would be to make `auth-utils` side-effect free by replacing `const SP = new SecurePassword()` with `const SP = () => newSecurePassword()`. That'd allow it do be tree-shaken correctly.

**Comment by flybayer** [comment]
Just making it side-effect free didn't work. Had to also add null-loader for secure-password.
